snapshots/snapshot_generator/src/main/scala/uk/ac/wellcome/platform/snapshot_generator/services/SnapshotService.scala

```diff
@@ -43,60 +51,100 @@ class SnapshotService(akkaS3Settings: S3Settings,
 
+    } yield
       CompletedSnapshotJob(
         snapshotJob = snapshotJob,
-        targetLocation = targetLocation
+        snapshotResult = snapshotResult
       )
-    }
   }
 
-  private def runStream(publicBucketName: String,
-                        publicObjectKey: String,
-                        index: Index,
-                        toDisplayWork: Work.Visible[Identified] => DisplayWork)
-    : Future[MultipartUploadResult] = {
+  private def getObjectMetadata(
+    location: S3ObjectLocation): Future[ObjectMetadata] = {
+    val objectMetadataFlow = S3
+      .getObjectMetadata(bucket = location.bucket, key = location.key)
+      .withAttributes(S3Attributes.settings(akkaS3Settings))
+      .map {
+        case Some(objectMetadata) => objectMetadata
+        case None =>
+          val runtimeException = new RuntimeException(
+            s"No object found at $location"
+          )
+
+          error(
+            msg = "Failed getting ObjectMetadata!",
+            t = runtimeException
+          )
+
+          throw runtimeException
+      }
+
+    objectMetadataFlow.runWith(Sink.head)
+  }
+
+  private def uploadSnapshot(
+    location: S3ObjectLocation,
+    index: Index,
+    toDisplayWork: Work.Visible[Identified] => DisplayWork)
+    : (Future[Int], Future[MultipartUploadResult]) = {
 
-    // This source outputs DisplayWorks in the elasticsearch index.
     val displayWorks: Source[DisplayWork, Any] =
       ElasticsearchWorksSource(elasticClient = elasticClient, index = index)
         .via(IdentifiedWorkToVisibleDisplayWork(toDisplayWork))
 
-    // This source generates JSON strings of DisplayWork instances, which
-    // should be written to the destination snapshot.
-    val jsonStrings: Source[String, Any] = displayWorks
-      .via(DisplayWorkToJsonStringFlow.flow)
-
-    // This source generates gzip-compressed JSON strings, corresponding to
-    // the DisplayWork instances from the source snapshot.
-    val gzipContent: Source[ByteString, Any] = jsonStrings
-      .via(StringToGzipFlow())
-
     val s3Sink: Sink[ByteString, Future[MultipartUploadResult]] =
       S3.multipartUpload(
-          bucket = publicBucketName,
-          key = publicObjectKey
+          bucket = location.bucket,
+          key = location.key
         )
         .withAttributes(S3Attributes.settings(akkaS3Settings))
 
-    gzipContent.runWith(s3Sink)
+    val countingSink: Sink[DisplayWork, Future[Int]] =
+      Sink.fold[Int, DisplayWork](0)((acc, _) => acc + 1)
+
+    val graph =
+      RunnableGraph.fromGraph(GraphDSL.create(countingSink, s3Sink)((_, _)) {
+        implicit builder => (countingSinkShape, s3SinkShape) =>
+          import GraphDSL.Implicits._
+
+          val broadcast = builder.add(Broadcast[DisplayWork](2))
+
+          displayWorks ~> broadcast.in
+          broadcast.out(0) ~> countingSinkShape
+          broadcast
+            .out(1) ~> DisplayWorkToJsonStringFlow.flow ~> StringToGzipFlow() ~> s3SinkShape
```

Comment:
Can you do something like
```scala
val workSnapshotSink = DisplayWorkToJsonStringFlow.flow ~> StringToGzipFlow() ~> s3SinkShape
...
broadcast.out(1) ~> workSnapshotSink
```
just to highlight that actually not all that much is happening here? I know it's a nit but the graph DSL is bananas and we should do all we can to help make it readable

Reply:
I would go one further, and suggest we isolate this Graph DSL logic in its own file (similar to the other flows in the snapshot generator). Something like:

```scala
object CountingSink
  def create[In, Out](
    source: Source[In, Any],
    sink: Sink[In, Result]): RunnableGraph[(Future[Int], Future[Result])]
}
```

and then you can test all this graph DSL logic without tying it to the SnapshotService (and we can reuse it later if we want to).

Reply:
> val snapshotSinkShape = DisplayWorkToJsonStringFlow.flow ~> StringToGzipFlow() ~> s3SinkShape

Is not valid syntax (surprisingly).

Reply:
I have attempted to reformat the way the graph DSL is used and some comments to clarify what the graph is doing.